Add docs on 'Lift and Shift' migration strategy (#17305)
cockroachdb/docs: src/current/v23.1/migration-strategy-lift-and-shift.md

---
title: "Migration Strategy: Lift and Shift"
summary: Learn about the 'Lift and Shift' data migration strategy
toc: true
docs_area: migrate
---

There are multiple strategies for [migrating off legacy technology](migration-overview.html#molt) to CockroachDB.

> This page discusses the 'Lift and Shift' strategy for migrating your data, which is a commonly used approach. This approach, which is also known as 'Big Bang' (and by other names), refers to the process where your data is moved in its entirety from a source system to a target system within a defined period of time. This can involve some application downtime or service degradation.

Author: Updated, thanks!  

(PS I appreciate the suggestion via the Github UI tool, just wanted to let you know I'm adding them manually. I don't accept them via the web interface since I rebase locally and it's hard to manage keeping my local edits in sync with the weird commits the UI makes. Just in case it seems strange that I'm saying "yes thanks" but not clicking the button here)

> Lift and Shift is a commonly used approach. However, it may not be the right approach if a strong application service continuity during the migration is required. It may be a viable method if application downtime is permitted.

Reviewer: You already say commonly used in the preceding paragraph.

I'd suggest deleting and starting this paragraph with "Lift and shift may not be the right approach..."

Reviewer 2: +1

Author: Done - thanks

{{site.data.alerts.callout_info}}
The information on this page assumes you have already reviewed the [Migration Overview](migration-overview.html).
{{site.data.alerts.end}}

## Pros and Cons

> On the spectrum of different data migration strategies, 'lift and shift' has the following pros and cons. The terms "lower" and "higher" are not absolute, but relative to other approaches.

Reviewer 2: ```suggestion
On the spectrum of different data migration strategies, Lift and Shift has the following pros and cons. The terms "lower" and "higher" are not absolute, but relative to other approaches.
```

Author: Done, thanks

Pros:

> - Conceptually straightforward

Reviewer 2: ```suggestion
- Conceptually straightforward.
```

Author: ✅

> - Less complex: if you can afford some downtime, the overall effort will usually be lower, and the chance of errors is lower

Reviewer 2: ```suggestion
- Less complex: If you can afford some downtime, the overall effort will usually be lower, and the chance of errors is lower.
```

Author: ✅

> - Lower time start-to-finish: in general, the more downtime you can afford, the shorter the overall migration project timeframe can be

Reviewer 2: ```suggestion
- Lower time start-to-finish: In general, the more downtime you can afford, the shorter the overall migration project timeframe can be.
```

> - Lower technical risk: it does not involve running multiple systems alongside each other for an extended period of time

Reviewer 2: ```suggestion
- Lower technical risk: It does not involve running multiple systems alongside each other for an extended period of time.
```

Author: ✅

> - Easy to practice dry runs of import/export using testing/non-production systems

Reviewer 2: ```suggestion
- Easy to practice dry runs of import/export using testing/non-production systems.
```

Author: ✅

> - Good import/export tooling is available (e.g. external tools like: [AWS Database Migration Service (DMS)](aws-dms.html), [Qlik Replicate](qlik.html), [Striim](striim.html); or internal tools like [`IMPORT INTO`](import-into.html), [`COPY`](copy-from.html), [`cockroach userdata`](cockroach-userfile-upload.html))

Reviewer: super nit but i think its (e.g., ...)

Author: fixed, thanks

> - Good import/export tooling is available (e.g. external tools like: [AWS Database Migration Service (DMS)](aws-dms.html), [Qlik Replicate](qlik.html), [Striim](striim.html); or internal tools like [`IMPORT INTO`](import-into.html), [`COPY`](copy-from.html), [`cockroach userdata`](cockroach-userfile-upload.html))

Author: ✅

> - If your application already has regularly scheduled maintenance windows, your customers will not encounter application downtime

Reviewer 2: ```suggestion
- If your application already has regularly scheduled maintenance windows, your customers will not encounter application downtime.
```

Author: ✅

Cons:

- Higher project risk: It either works or does not work; once you start, you have to finish or rollback

> - Application service continuity requirements must be relaxed (that is, application downtime or degradation may be needed)

Reviewer: Do we need to say "service degradation" or "increased latency"

Author: increased latency sounds good - updated

> - Less control of the pace (the project *has* to be completed to meet a given downtime / service degradation window)

Reviewer: Maybe say "All or Nothing" instead of less control of the pace since i don't think pace itself is the concern but more the all or nothing. I think i'd name the first con all or nothing and then have the current test. I'd make a new bullet that says "high project risk" which communicates that there are fewer ways to mitigate migration complications during the  process"

Author: sounds good - renamed the one bullet to 'all or nothing' and edited the bullets as you suggested (if I understood correctly) - thanks!

> - Less control of the pace (the project *has* to be completed to meet a given downtime / service degradation window)

Author: Updated, thanks

## Process design considerations

{{site.data.alerts.callout_info}}

> The high-level considerations in this section only refer to the data loading portion of your migration. They assume you are following the steps in the overall migration process described in [Migrate your Database to CockroachDB](migration-overview.html).

Author: ✅

{{site.data.alerts.end}}

> Keep in mind the following considerations when designing a 'Lift and Shift' data migration process.

Author: Fixed

> - [Decide on your data migration tooling](#managed-migration)

Author: ✅

- [Decide which data formats you will use](#data-formats)
- [Design a restartable process](#restartable)

> - [Design a scalable process](#scalable)

Reviewer 2: ```suggestion
- [Design a scalable process](#scalable).
```

<a name="managed-migration"></a>

### Decide on your data migration tooling

> If you plan to do your bulk data migration using a managed migration service, you must have a secure, publicly available CockroachDB cluster. CockroachDB supports the following migration services:

Reviewer 2: ```suggestion
If you plan to do your bulk data migration using a managed migration service, you must have a secure, publicly available CockroachDB cluster. CockroachDB supports the following [third-party migration services](third-party-database-tools.html#data-migration-tools):
```

Author: oh nice, I did not know about that page - thanks!

> - [Migrate with AWS Database Migration Service (DMS)](aws-dms.html)

Reviewer 2: ```suggestion
- [AWS Database Migration Service (DMS)](aws-dms.html)
```

Author: Updated, thanks

> - [Migrate and Replicate Data with Qlik Replicate](qlik.html)

Reviewer 2: ```suggestion
- [Qlik Replicate](qlik.html)
```

Author: fixed

> - [Migrate and Replicate Data with Striim](striim.html)

Reviewer 2: ```suggestion
- [Striim](striim.html)
```

{{site.data.alerts.callout_info}}
Depending on the migration service you choose, [long-running transactions](query-behavior-troubleshooting.html#hanging-or-stuck-queries) can occur. In some cases, these queries will cause [transaction retry errors](common-errors.html#restart-transaction). If you encounter these errors while migrating to CockroachDB using a managed migration service, please reach out to our [Support Resources](support-resources.html).
{{site.data.alerts.end}}

> If you will not be using a managed migration service, see the following sections for more information on how to use SQL statements like [`IMPORT INTO`](import-into.html), [`COPY`](copy-from.html), etc.

Reviewer 2: ```suggestion
If you will not be using a managed migration service, see the following sections for more information on how to use SQL statements like [`IMPORT INTO`](import-into.html), [`COPY FROM`](copy-from.html), etc.
```

Author: fixed

<a name="data-formats"></a>

### Decide which data formats and storage media you will use

It's important to decide which data formats, storage media, and database features you will use to migrate your data.

Data formats that can be imported by CockroachDB include:

- [SQL](schema-design-overview.html) for the [schema import](../cockroachcloud/migrations-page.html).
- [CSV](migrate-from-csv.html) for table data.
- [Avro](migrate-from-avro.html) for table data.

The storage media you use to export / import from can be intermediate data files or streaming data coming over the network. Options include:

- Local "userdata" storage for small tables (see [`cockroach userdata`](cockroach-userfile-upload.html), [Use a Local File Server](use-a-local-file-server.html)).
- Cloud blob storage (see [`IMPORT INTO`](import-into.html), [Import Performance Best Practices](import-performance-best-practices.html), [Use Cloud Storage](use-cloud-storage.html)).
- Direct wire transfers (see [managed migration services](#managed-migration)).

Database features for export/import from the source and target databases can include:

> - Tools for exporting from the source database may include: `pg_dump --schema-only` and `COPY`, `mysqldump`, `expdp`, etc.

Reviewer 2: ```suggestion
- Tools for exporting from the legacy database may include: `pg_dump --schema-only` and `COPY FROM`, `mysqldump`, `expdp`, etc.
```

Author: oooh legacy now that's a loaded word :-) 
tbh that's getting into non-technical/marketing value judgement territory IMO so I'm gonna leave that one alone here, feel free to change it in another PR tho :-) 

PS I know I kinda already cracked the seal on the use of legacy in the MOLT reference at the top of this page due to the acronym but tbh I kinda feel bad about it because I think the word "legacy" is always used by new things and codes as somewhat insulting to reliable existing technologies and I don't usually like to use it. In fact I often feel a knee-jerk skepticism whenever some "new hotness" thing starts calling all the things that already exist "legacy", especially when those "legacy" things actually have, y'know, way more users and market share and track record :-) 

after writing this I feel my MOLT ref in the intro to this page is too cute by half - I'd kinda like to change it but I don't want to take the time to re-litigate what to say in its place at this point

def updated the `COPY FROM` tho

> - For import into CockroachDB, use [`COPY`](copy-from.html) or [`IMPORT INTO`](import-into.html). For a bulk data migrations, most users should use `IMPORT INTO` because the tables will be offline anyway, and `IMPORT INTO` can [perform the data import much faster](import-performance-best-practices.html) than `COPY`.

Reviewer 2: ```suggestion
- For import into CockroachDB, use [`COPY FROM`](copy-from.html) or [`IMPORT INTO`](import-into.html). For a bulk data migrations, most users should use `IMPORT INTO` because the tables will be offline anyway, and `IMPORT INTO` can [perform the data import much faster](import-performance-best-practices.html) than `COPY FROM`.
```

Author: Fixed

Decide which of the options above will meet your requirements while resulting in a process that is [restartable](#restartable) and [scalable](#scalable).

<a name="restartable"></a>

### Design a restartable process

To have a higher chance of success, design your data migration process so it can be stopped and restarted from an intermediate state at any time during the process. This will help minimize errors and avoid wasted effort.

Keep the following requirements in mind as you design a restartable import/export process:

- Bulk migrate data in manageable size batches for your source and target systems.
    - This is a best practice. If something happens to the target cluster during import, the amount of wasted work will be minimized.
- Implement progress/state keeping with process restart capabilities.
- Make sure your export process is idempotent: the same input to your export process should return the same output data.

> - If possible, export and import the majority of your data before taking the source database down. This can ensure that you only have to deal with the incremental changes from your last import to complete the migration process.

Reviewer 2: ```suggestion
- If possible, export and import the majority of your data before taking down the legacy database. This can ensure that you only have to deal with the incremental changes from your last import to complete the migration process.
```

Author: took the word order rephrasing, leaving legacy alone for now as mentioned in other comment. :-)

<a name="scalable"></a>

### Design a scalable and performant process

Once your process is [restartable and resilient to failures](#design-a-restartable-process), it's important to also make sure it will scale to the needs of your data set. The larger the data set you are migrating to CockroachDB, the more important the performance and scalability of your process will be.

> Keep the following requirements in mind:

Reviewer 3: Could be useful to note in here that benchmarking performance will help ensure the process completes within the allotted downtime window.

Author: thanks! updated to add another bullet that says

> - Benchmark the performance of your migration process to help ensure it will complete within the allotted downtime window.

- Schema and data should be imported separately.
- Your process should handle multiple files across multiple export/import streams concurrently.

>   - For best performance, these files should contain presorted, disjoint data sets.

Reviewer 2: Should that be "disjointed" or is this a term I don't know?

Author: ya it's a sets term: https://en.wikipedia.org/wiki/Disjoint_sets

---
title: "Migration Strategy: Lift and Shift"
summary: Learn about the 'Lift and Shift' data migration strategy
toc: true
docs_area: migrate
---

There are multiple strategies for [migrating off legacy technology](migration-overview.html#molt) to CockroachDB.

> This page discusses the 'Lift and Shift' strategy for migrating your data, which is the most commonly used approach. This may not be the right approach if a strong application service continuity during the migration is required. However, it may be a viable method if downtime is permitted. It works particularly well when the application accurately maintains [primary key](primary-key.html) and [foreign key](foreign-key.html) constraints.

Reviewer: I think we should slightly hedge the most commonly used to something like "which is a popular approach"

Should we say application downtime?

Also, why does it work well for PK and FK constraints?

Author: > I think we should slightly hedge the most commonly used to something like "which is a popular approach"

updated to say "a commonly used approach"

> Should we say application downtime?

updated mentions of "downtime" to clarify "application downtime"

> Also, why does it work well for PK and FK constraints?

this was based on some internal writings. we could probably get into it for a v2 but for now I've deleted the PK / FK bit based on some feedback via https://cockroachlabs.atlassian.net/wiki/x/IgAgtw from @sheaffej and @a-entin

## Pros and Cons

On the spectrum of different data migration strategies, 'lift and shift' has the following pros and cons. The terms "low" and "high" are not absolute, but relative to other approaches.

Pros:

> 

Reviewer: We should add something like if your application already has regularly scheduled maintenance windows no customers will encounter application downtime

Author: Nice, good point.  Added that!

- Low effort: in general, if you can afford a bit of downtime, the overall effort will be lower

> - Low cost: in general, "lift and shift" costs less than having your development team modify an application to write to two data sources, or run some kind of syncing between source and target systems

Reviewer: What kind of costs? FTE time? Hardware? Oppportunity cost?  Can we be more explicit?

Author: I'm inclined to delete this point since we don't have hard numbers to support.  For now I've deleted it

in general one can imagine that monetary costs would be lower than maintaining / updating an app that is doing the "strangler" approach as outlined in https://blog.cloudneutral.se/effective-strategies-for-planning-and-executing-data-migration-projects 

but we may have to get into that in a v2 of this doc since it's very hard to quantify in a generic overview doc like this, I think

> - Low time start-to-finish: in general, the more downtime you can afford, the shorter the overall migration timeframe can be

Reviewer: Can we work in here its the overall _project_ that is faster?

Author: Sure, updated to say "the shorter the overall migration project timeframe can be" - do you think that is sufficient?

- Low technical risk: it does not involve running multiple systems alongside each other for an extended period of time

> - Good import/export tooling is available

Reviewer: Should we link to the internal and external tools here?

Author: Sounds good - added a list of internal and external tools to this bullet

- Easy to practice dry runs of import/export using testing/non-production systems
- Conceptually straightforward

Cons:

> - High project risk: it involves at least some downtime and it either works or does not work

Reviewer: I think we should split these into two rows
1. Incurs application downtime
2. High project risk

Author: Sounds good - broken into separate bullets

- Application service continuity requirements must be relaxed (i.e., some kind of application downtime or degradation is usually needed)
- Less control of the pace (*has* to be completed to meet a given downtime / service degradation window)

## Process design considerations

{{site.data.alerts.callout_info}}
The high-level considerations in this section only refer to the data loading portion of your migration. They assume you are following the steps in the overall migration process described in [Migrate your Database to CockroachDB](migration-overview.html).
{{site.data.alerts.end}}

Keep in mind the following considerations when designing a 'Lift and Shift' data migration process.

- [Decide how much downtime you can tolerate](#downtime)

> - [Decide if you will use a managed migration service](#managed-migration)

Reviewer: Should this be retitled to, Decide your data migration tooling? (it seems more immediately obvious what to decide on)

Author: Sounds good - I have updated to use that section title, and added some text in that section directing users to see the other sections for more info about `IMPORT INTO` etc. if they are not using a managed service

- [Decide which data formats you will use](#data-formats)
- [Design a restartable process](#restartable)
- [Design a scalable process](#scalable)

<a name="downtime"></a>

> ### Decide how much downtime you can tolerate

Reviewer 2: Since exact topic is covered [in the new `migration-overview.md`](https://github.com/cockroachdb/docs/pull/17339) (see preview: https://deploy-preview-17339--cockroachdb-docs.netlify.app/docs/v23.1/migration-overview.html#approach-to-downtime), and I assume this doc should be published after that one is finalized, I think it would be best (most consistent) to use an include across both docs instead of the below.

@rmloveland Maybe we can sync separately about the contents now that they've become clearer in the migration overview. Sorry that working in parallel is requiring some back-and-forth.

Author: > Since exact topic is covered [in the new migration-overview.md](https://github.com/cockroachdb/docs/pull/17339) (see preview: https://deploy-preview-17339--cockroachdb-docs.netlify.app/docs/v23.1/migration-overview.html#approach-to-downtime), and I assume this doc should be published after that one is finalized, I think it would be best (most consistent) to use an include across both docs instead of the below.

As Ryan and I just discussed on a call, for v1 of this doc we're going to remove this section from this doc and mention on this page that we expect users will be familiar with the Migration Overview doc 

> @rmloveland Maybe we can sync separately about the contents now that they've become clearer in the migration overview. Sorry that working in parallel is requiring some back-and-forth.

No worries!

{{site.data.alerts.end}}

<a name="data-formats"></a>

### Decide which data formats and storage media you will use

It's important to decide which data formats, storage media, and database features you will use to migrate your data.

Data formats that can be imported by CockroachDB include:

- [SQL](schema-design-overview.html) for the [schema import](../cockroachcloud/migrations-page.html).
- [CSV](migrate-from-csv.html) for table data.
- [Avro](migrate-from-avro.html) for table data.

The storage media you use to export / import from can be intermediate data files or streaming data coming over the network. Options include:

- Local "userdata" storage (see [`cockroach userdata`](cockroach-userfile-upload.html), [Use a Local File Server](use-a-local-file-server.html)).
- Cloud blob storage (see [`IMPORT INTO`](import-into.html), [Import Performance Best Practices](import-performance-best-practices.html), [Use Cloud Storage](use-cloud-storage.html)).

> - Direct wire transfers (see [github.com/cockroachlabs/cdc-sink](https://github.com/cockroachlabs/cdc-sink) or [managed migration services](#managed-migration)).

Reviewer 2: @awoods187 I think we should avoid talking about CDC for migration use cases for now, right?

Reviewer: For ingestion, it should be import/copy, aws dms, qlik, and striim, and i guess technically inserts/upserts. I think we should not include CDC sink until we decided separately.

Author: thanks, I've deleted all references to cdc-sink in this doc for now

Database features for export/import from the source and target databases can include:

- Tools for exporting from the source database may include: `pg_dump --schema-only` and `COPY`, `mysqldump`, `expdp`, etc.
- For import into CockroachDB, use [`COPY`](copy-from.html) or [`IMPORT INTO`](import-into.html). For a bulk data migrations, most users should use `IMPORT INTO` because the tables will be offline anyway, and `IMPORT INTO` can [perform the data import much faster](import-performance-best-practices.html) than `COPY`.

Decide which of the options above will meet your requirements while resulting in a process that is [restartable](#restartable) and [scalable](#scalable).

<a name="restartable"></a>

### Design a restartable process

To have a higher chance of success, design your data migration process so it can be stopped and restarted from an intermediate state at any time during the process. This will help minimize errors and avoid wasted effort.

Keep the following requirements in mind as you design a restartable import/export process:

> 

Reviewer: Somewhere we should talk about the fact that you can export and import the majority of your data before taking your source db down to drain such that you only have to deal with the incremental changes from your last import

Author: nice, thanks

added 

> - If possible, export and import the majority of your data before taking the source database down. This can ensure that you only have to deal with the incremental changes from your last import to complete the migration process.
